src/media-utils.ts

```diff
@@ -154,9 +180,108 @@ function getVideoConstraints (cameraOnly: boolean): MediaStreamConstraints {
     };
   }
 
-  if (cameraOnly) {
+  if (!opts.video) {
+    constraints.video = false;
+  }
+
+  if (!opts.audio) {
     constraints.audio = false;
   }
 
   return constraints;
 }
+
+let refreshDevices = true;
+const enumeratedDevices: IEnumeratedDevices = {
+  videoDeviceIds: [],
+  audioDeviceIds: [],
+  outputDeviceIds: []
+};
+
+async function getEnumeratedDevices (sdk: PureCloudWebrtcSdk): Promise<IEnumeratedDevices> {
+  if (!window.navigator.mediaDevices || !window.navigator.mediaDevices.enumerateDevices) {
+    log.call(sdk, LogLevels.warn, 'Unable to enumerate devices');
+    return enumeratedDevices;
+  }
+
+  if (!window.navigator.mediaDevices.ondevicechange) {
+    window.navigator.mediaDevices.ondevicechange = () => {
+      log.call(sdk, LogLevels.debug, 'onDeviceChange fired');
+      refreshDevices = true;
+    };
+  }
+
+  // if devices haven't changed since last time we called this
+  if (!refreshDevices) {
+    return enumeratedDevices;
+  }
+
+  try {
+    enumeratedDevices.videoDeviceIds = [];
+    enumeratedDevices.audioDeviceIds = [];
+    enumeratedDevices.outputDeviceIds = [];
+
+    const devices = await window.navigator.mediaDevices.enumerateDevices();
+    refreshDevices = false;
+    devices.forEach(device => {
+      if (device.kind === 'videoinput') {
+        enumeratedDevices.videoDeviceIds.push(device.deviceId);
+      } else if (device.kind === 'audioinput') {
+        enumeratedDevices.audioDeviceIds.push(device.deviceId);
+      } else if (device.kind === 'audiooutput') {
+        enumeratedDevices.outputDeviceIds.push(device.deviceId);
+      }
+    });
+  } catch (e) {
+    throwSdkError.call(sdk, SdkErrorTypes.generic, 'Error enumerating devices', e)
+  }
+
+  return enumeratedDevices;
+}
+
+export async function getDeviceByKindAndId (sdk: PureCloudWebrtcSdk, kind: MediaDeviceKind, deviceId: boolean | string): Promise<string | undefined> {
+
+  if (deviceId === false) {
```

Comment:
we should probably throw if deviceId is a boolean. I think that will just make things easier. there's no reason to search devices if it's a boolean.